### PR TITLE
Resurrect the OpenBSD Build

### PR DESCRIPTION
**Pull request by mikhail-diakov**
Major issues addressed:
  - GNU extensions in build scripts: using gtar, gmake from ports.
  - OpenBSD lacks Unified Buffer Cache, therefore LMDB environment should be opened with MDB_WRITEMAP flag.
  - Lack of custom memory allocators support in LibreSSL, so building OpenSSL.

FYI, default  `datasize` limit should be increased to successfully run Urbit.

### CODE REVIEW

**Comment by mrdomino** [comment]
Tried to submit a PR against this one, but I was unable to find your fork in the UI just now. I merged `next/vere` into this branch and included a couple fixes here:

https://github.com/urbit/urbit/tree/jo/obsd-next-vere

(N.B. @mikhail-diakov I probably should have advised you to base this off of `next/vere` in the first place; it's generally correct to base work off of `master` and then merge to the appropriate release branch later as we're doing here, but in this case some parallel work happened on `next/vere` in the meantime.)

**Comment by mrdomino** [comment]
Aside from the proposed changes in `jo/obsd-next-vere`, this looks ready to go.

**Comment by joemfb** [comment]
@mrdomino here's the link for comparing/PR'ing: https://github.com/mikhail-diakov/urbit/compare/openbsd-build...urbit:/jo/obsd-next-vere

**Comment by mikhail-diakov** [comment]
Merged `obsd-next-vere`, give me some time to test the current state.

File: pkg/landscape/app/graph-store.hoon [hunk 1]
```diff
@@ -860,8 +860,8 @@
       --
     ::
         [%depth-first @ @ ~]
-      =/  count=(unit atom)  (rush i.t.t.t.t.path dem:ag)
-      =/  start=(unit atom)  (rush i.t.t.t.t.t.path dem:ag)
+      =/  count=(unit atom)  (rush i.t.t.t.t.t.path dem:ag)
+      =/  start=(unit atom)  (rush i.t.t.t.t.t.t.path dem:ag)
```
Comment (mikhail-diakov) [inline pkg/landscape/app/graph-store.hoon:864]: Are these changes intentional?

**Comment by joemfb** [comment]
Sorry for the delay @mikhail-diakov, these changes all look good.

I'm not sure how the extra changes ended up on this branch, I'm guessing that something went wrong in @mrdomino's merge. I just tried a local merge of your commit (ad7712e624dcc9081fa1c20e2ac2b30e091c7d9c), and the only conflict is trivial (renaming `make`/`gmake`/`gnumake` in the "poor man's nix-shell". So I think it'd be easiest for you to rebase out the merge commits -- I can handle the conflict on merge, or you can merge `next/vere` back into your branch (or just rebase on top of it). 

For future testing/maintenance, can you recommend a source for cloud-based openbsd vms, or a virtualization mechanism? Ideally we could get it into CI, but it looks like that would be quite difficult.

File: pkg/landscape/app/graph-store.hoon (hunk 1 above)
Comment (mikhail-diakov) [reply pkg/landscape/app/graph-store.hoon]: Not relevant anymore.

**Comment by mikhail-diakov** [comment]
Removed merge commit, rebased onto `next/vere`, resolved conflicts, and fixed build.

**Comment by mikhail-diakov** [comment]
# Building urbit on OpenBSD
This guide assumes fresh OpenBSD 7.0 installation.

### Basic setup
Make sure your user has doas privileges.
```
su
echo "permit persist :wheel" > /etc/doas.conf
usermod -G wheel <my_user>
exit
```

Sync your watches.
```
doas rdate http://pool.ntp.org
```

Install prerequisite dependencies.
```
doas pkg_add bash git git-lfs automake-1.16.3 libtool jq gtar-1.34 cmake gmake libsigsegv gmp
```

### Cloning the repo
```
git clone <target_origin> <clone_dir>
cd <clone_dir>
git checkout -b <target_branch>
git-lfs install
git-lfs pull
```

### Building urbit
```
cd <clone_dir>/pkg/urbit
./configure
gmake build/urbit
```

Resulting binary can be found in `<clone_dir>/pkg/urbit/build/`.

## Running urbit
Urbit requires more memory for its process than allowed by default on OpenBSD. One of the ways to fix this is as follows.
  - `doas vim /etc/login.conf`
  - Ensure that `staff` group has `infinity` set for it's `datasize-cur` and `datasize-max` properties. It should look something like this.
```
staff:\
        :datasize-cur=infinity:\
        :datasize-max=infinity:\
        :maxproc-max=512:\
        :maxproc-cur=256:\
        :ignorenologin:\
        :requirehome@:\
        :tc=default:
```
  - `doas cap_mkdb /etc/login.conf`
  - `doas usermod -G staff <my_user>`
  - Exit and re-login into the shell.

You can now run urbit the usual way, e.g. `./urbit -c my-obsd-test-comet`.

**Comment by nartesfasrum** [comment]
Some quick notes from a fairly naive run through the instructions on a fresh OpenBSD 7.1 install:

- `doas rdate http://pool.ntp.org` seems not to work.
- `./configure` throws an "Unsupported or unknown CPU type: amd64" error on my 64-bit VM. 
- `./configure` throws an error about not finding `libh2o` and `liburcrypt-0` in the `pkg-config` search path

**Comment by mikhail-diakov** [comment]
What branch are you testing the build on?

I've supported amd64 CPU type, here is the relevant diff: https://github.com/urbit/urbit/pull/5709/files#diff-27310e396648c214f35ce3ee0262c4ce02ab4a3ab565a360976d4b752255c23aR88

**Comment by mikhail-diakov** [comment]
You're right regarding `rdate`, it should be `doas rdate pool.ntp.org` (without `http://`). If `pool.ntp.org` isn't working, please try another time server (for instance, `utcnist.colorado.edu`).

Updated the instructions.

**Review by joemfb — APPROVED** [review]
Excellent! Thank you @mikhail-diakov and @nartesfasrum.
